### PR TITLE
Feature/17176 Render pupil scores on pupil register page

### PR DESCRIPTION
**Pull request by GuyHarwood**
Interim work on surfacing the check results as a percentage.  Specifically for Pauls demo on Thursday.

The marking is done when the check is received from the SPA.  This will eventually be replaced by a job that will continually process results in Azure.

### CODE REVIEW

File: admin/controllers/school.js [hunk 1]
```diff
@@ -9,11 +9,7 @@ const ValidationError = require('../lib/validation-error')
 const errorConverter = require('../lib/error-converter')
 const hdfErrorMessages = require('../lib/errors/hdf')
 const hdfValidator = require('../lib/validator/hdf-validator')
-const {
-  fetchAnswers,
-  fetchScoreDetails,
-  fetchMultiplePupils,
-  fetchOnePupil } = require('../services/pupil.service')
+const pupilService = require('../services/pupil.service')
```
Comment (GuyHarwood) [inline admin/controllers/school.js]: @pris54 @jon-shipley @GeorgeChatzigiannis breaking down a service into function variables is not nice to work with.  Can we please ensure we require the entire service as a variable going forward.

File: admin/controllers/school.js [hunk 2]
```diff
@@ -22,6 +18,7 @@ const schoolDataService = require('../services/data-access/school.data.service')
 const generatePinsService = require('../services/generate-pins.service')
 const { sortRecords } = require('../utils')
 const sortingAttributesService = require('../services/sorting-attributes.service')
+const checkDataService = require('../services/data-access/check.data.service')
```
Comment (GuyHarwood) [inline admin/controllers/school.js]: @pris54 @GeorgeChatzigiannis  This controller depends upon services **and** data-services.  It uses the data-services to pass data to the services!!!  This is a dependency nightmare.  We need to refactor this out going forward.

File: admin/controllers/school.js (hunk 1 above)
Comment (GuyHarwood) [reply admin/controllers/school.js]: Also, this school controller is insanely long.  It needs breaking down in a big way.  I can see at least 2 other controllers such as results and attendance that could be made here.

**Comment by GuyHarwood** [comment]
@pris54 @sinhaamit1980 @activemq Travis is reporting failing scenarios that are unrelated to my feature...

```Failing Scenarios:
cucumber features/add_multiple_pupil.feature:7 # Scenario: Add Multiple Pupil Landing page displays heading and sub heading section
cucumber features/create_check_window.feature:20 # Scenario: Users can create a check window
cucumber features/edit_check_window.feature:6 # Scenario: Edit check window displays pre populated data
cucumber features/edit_check_window.feature:9 # Scenario: Check Window detail is updated when valid details are entered
cucumber features/edit_check_window.feature:15 # Scenario: User can Navigate back to Manage Check Window page
cucumber features/manage_check_window.feature:43 # Scenario: Modal has a heading
cucumber features/manage_check_window.feature:47 # Scenario: Modal has content
cucumber features/manage_check_window.feature:51 # Scenario: Modal has a option to cancel
cucumber features/manage_check_window.feature:55 # Scenario: Modal has a option to confirm
cucumber features/manage_check_window.feature:60 # Scenario: Windows that have a remove button can be removed
cucumber features/manage_check_window.feature:66 # Scenario: Removal of check window can be cancelled```

File: admin/controllers/school.js [hunk 3]
```diff
@@ -97,13 +106,12 @@ const getResults = async (req, res, next) => {
   const { pupils, schoolData } = await pupilDataService.getPupils(req.user.School)
   let pupilsFormatted = await Promise.all(pupils.map(async (p) => {
     const fullName = `${p.foreName} ${p.lastName}`
-    const answers = await fetchAnswers(p._id)
-    const { hasScore, score, percentage } = fetchScoreDetails(answers)
+    let score = await getScorePercentage(p._id)
```
Comment (GeorgeChatzigiannis) [inline admin/controllers/school.js]: why let instead of const here?

File: admin/controllers/school.js [hunk 4]
```diff
@@ -268,14 +276,15 @@ const getSubmitAttendance = async (req, res, next) => {
   let pupilsFormatted = await Promise.all(pupils.map(async (p) => {
     const fullName = `${p.foreName} ${p.lastName}`
     const { _id: id, hasAttended } = p
-    const answers = await fetchAnswers(p._id)
-    const { hasScore, percentage } = fetchScoreDetails(answers)
+    let score = await getScorePercentage(p._id)
```
Comment (GeorgeChatzigiannis) [inline admin/controllers/school.js]: same comment as above

File: admin/controllers/school.js [hunk 5]
```diff
@@ -407,7 +416,7 @@ const getPupilNotTakingCheck = async (req, res, next) => {
   // Flash message after removing a reason
   if (req.params.removed) {
     let flashMessage = 'Reason removed'
-    let pupil = await fetchOnePupil(req.params.removed, req.user.School)
+    let pupil = await pupilService.fetchOnePupil(req.params.removed, req.user.School)
```
Comment (GeorgeChatzigiannis) [inline admin/controllers/school.js]: seems like that should be a const too

File: admin/services/data-access/check.data.service.js [hunk 6]
```diff
@@ -22,6 +22,15 @@ checkDataService.find = async function (criteria) {
 }
 
 /**
+ * Find Checks by criteria: e.g. checkDataService.find({checkWindowId: 1234})
+ * @param criteria
+ * @return {Promise.<void>} - lean Check objects
+ */
+checkDataService.findLatestCheckByPupilId = async function (pupilId) {
+  return Check.findOne({ pupilId: pupilId }).sort({ field: 'asc', _id: -1 }).limit(1).lean().exec()
```
Comment (jon-shipley) [inline admin/services/data-access/check.data.service.js]: `findOne() *and* `limit(1)` seems like belt and braces?

File: admin/controllers/school.js [hunk 7]
```diff
@@ -42,6 +39,19 @@ const getHome = async (req, res, next) => {
   })
 }
 
+const getScorePercentage = async (pupilId) => {
```
Comment (GeorgeChatzigiannis) [inline admin/controllers/school.js]: shouldn't that be within a service instead of the controller?

Comment (GeorgeChatzigiannis) [reply admin/controllers/school.js]: Also it will need to be tested

File: admin/services/pupil.service.js [hunk 8]
```diff
@@ -41,19 +39,24 @@ const pupilService = {
...
-  fetchScoreDetails: (answers) => {
-    const pupilScore = answers && answers.result
-    const hasScore = !!pupilScore && typeof pupilScore.correct === 'number' && pupilScore.correct >= 0
-    const score = pupilScore ? `${pupilScore.correct}/${answers.answers.length}` : 'N/A'
-    const percentage = pupilScore ? `${Math.round((pupilScore.correct / answers.answers.length) * 100)}%` : 'N/A'
-    return {
-      hasScore,
-      score,
-      percentage
+  calculateScorePercentage: (results) => {
```
Comment (GeorgeChatzigiannis) [inline admin/services/pupil.service.js]: Need tests for these one as well

File: admin/services/data-access/check.data.service.js (hunk 6 above)
Comment (GuyHarwood) [reply admin/services/data-access/check.data.service.js]: yep, superfluous!

File: admin/controllers/school.js (hunk 7 above)
Comment (GuyHarwood) [reply admin/controllers/school.js]: This is a helper function that I extracted from repeated code within the controller.

The school controller needs a complete overhaul (622 lines 😱), so I would rather it be done as one job. 
This needs to go out today for Pauls demo tomorrow.  
I will create a PBI for the controller refactoring.

Comment (GeorgeChatzigiannis) [reply admin/controllers/school.js]: Indeed but we still to only contain pure controller methods every single helper function should exist to a service imo(like a utilService etc.)
Best to keep a note for the refactoring PBI.

**Pull request merged**